Add ERC: Confidential Fungible Token (#1113)
ethereum/ERCs: ERCS/erc-7984.md

### Nomenclature

- All amounts in this ERC are pointer based amounts represented by `bytes32` pointers unless otherwise specified. The resolution and manipulation of these pointers is implementation specific.

### Token

### Methods

Compliant tokens MUST implement the following methods, unless otherwise specified:

#### `name()`

Returns the name of the token - e.g. `"MyConfidentialToken"`.

OPTIONAL - This method can be used to improve usability, but interfaces and other contracts MUST NOT expect these values to be present.

```solidity
function name() external view returns (string memory)
```

#### `symbol()`

Returns the symbol of the token - e.g. `"MCT"`.

OPTIONAL - This method can be used to improve usability, but interfaces and other contracts MUST NOT expect these values to be present.

```solidity
function symbol() external view returns (string memory)
```

#### `decimals()`

> Returns the number of decimals the token uses (e.g. `6`) as a plaintext `uint8`.

Reviewer: Should this be marked optional ?

```solidity
function onConfidentialTokenReceived(address from, bytes32 amount, bytes calldata data) external returns (bytes32 success);
```

## Rationale

### Technology Agnostic Design

Using `bytes32` allows implementations using pointer based systems and privacy mechanisms including FHE systems, zero-knowledge proofs, secure enclaves, or future technologies to be compliant.

### Operator Model

Time-limited operators provide granular control while enabling DeFi protocol integration and natural permission expiration. This approach reduces the load on the external system by removing the need to track approval amounts.

### Data Parameter

The `bytes calldata data` parameter in transfer functions allows implementations to include cryptographic proofs, access permissions, or other privacy-mechanism-specific information.

## Security Considerations

Security depends on the underlying pointer based mechanism. Implementations must guard against side-channel attacks and ensure proper key management for offchain operations.

> 

Author: ```suggestion

Token callbacks are associated with inherent security issues, including reentrancy risks and gas griefing. When utilizing callbacks, consider using reentrancy protection and setting a gas limit. 
```

---
eip: 7984
title: Confidential Fungible Token Interface
description: An interface for confidential fungible tokens
author: Aryeh Greenberg (@arr00), Ernesto García (@ernestognw), Hadrien Croubois (@Amxx), Ghazi Ben Amor (@GBAZama), Clement Danjou (@immortal-tofu), Joseph Andre Turk (@jatZama), Silas Davis (@silasdavis), Nicolas Pasquier (@npasquie)

> discussions-to: https://ethereum-magicians.org/t/new-erc-confidential-fungible-token-standard/24735

Reviewer 2: ```suggestion
discussions-to: https://ethereum-magicians.org/t/erc-7984-confidential-fungible-token-interface/24735
```
Add assigned ERC and title

### Nomenclature

- All amounts in this ERC are pointer based amounts represented by `bytes32` pointers unless otherwise specified. The resolution and manipulation of these pointers is implementation specific.

### Token

### Methods

Compliant tokens MUST implement the following methods, unless otherwise specified:

#### `name()`

Returns the name of the token - e.g. `"MyConfidentialToken"`.

OPTIONAL - This method can be used to improve usability, but interfaces and other contracts MUST NOT expect these values to be present.

```solidity
function name() external view returns (string memory)
```

#### `symbol()`

Returns the symbol of the token - e.g. `"MCT"`.

OPTIONAL - This method can be used to improve usability, but interfaces and other contracts MUST NOT expect these values to be present.

```solidity
function symbol() external view returns (string memory)
```

#### `decimals()`

Returns the number of decimals the token uses (e.g. `6`) as a plaintext `uint8`.

```solidity
function decimals() external view returns (uint8)
```

#### `confidentialTotalSupply()`

> Returns the total token supply.

Reviewer 3: 1. Which actor/component would legitimately be able to view total supply?
2. If we have one, maybe we should add in "Security Considerations" a warning claiming implementation should prevent deducting balances by querying total supply frequently

### Nomenclature

- All amounts in this ERC are pointer based amounts represented by `bytes32` pointers unless otherwise specified. The resolution and manipulation of these pointers is implementation specific.

### Token

### Methods

Compliant tokens MUST implement the following methods, unless otherwise specified:

#### `name()`

Returns the name of the token - e.g. `"MyConfidentialToken"`.

OPTIONAL - This method can be used to improve usability, but interfaces and other contracts MUST NOT expect these values to be present.

```solidity
function name() external view returns (string memory)
```

#### `symbol()`

Returns the symbol of the token - e.g. `"MCT"`.

OPTIONAL - This method can be used to improve usability, but interfaces and other contracts MUST NOT expect these values to be present.

```solidity
function symbol() external view returns (string memory)
```

#### `decimals()`

Returns the number of decimals the token uses (e.g. `6`) as a plaintext `uint8`.

```solidity
function decimals() external view returns (uint8)
```

#### `contractURI()`

> Returns the metadata URI for the token. SHOULD follow the schema defined in ERC7572.

Author: ```suggestion
Returns the metadata URI for the token. SHOULD follow the schema defined in ERC-7572.
```
